source/postprocess/visualization/density_anomaly.cc

```diff
@@ -0,0 +1,244 @@
+                    in.pressure[0]=this->get_adiabatic_conditions().pressure(in.position[0]);
+
+                    // Evaluate adiabatic/reference density at the cell center
+                    this->get_material_model().evaluate(in, adiabatic_out);
+                    const double adiabatic_density = adiabatic_out.densities[0];
+
+                    // Compute the percentage deviation from the average
+                    (*return_value.second)(cell->active_cell_index()) = density - adiabatic_density;
+                  }
+              break;
+            }
+            case lateral_average:
+            {
+              std::vector<double> density_depth_average(n_slices);
+              this->get_lateral_averaging().get_density_averages(density_depth_average);
+
+              // Estimates of the lateral density average at each depth are required
+              // for all cell depths, including those
+              // shallower than the midpoint of the first slice or
+              // deeper than the midpoint of the last slice.
+              std::vector<double> padded_density_depth_average(n_slices+2);
+
+              padded_density_depth_average[0] = 2.*density_depth_average[0] - density_depth_average[1];
+              padded_density_depth_average[n_slices+1] = 2.*density_depth_average[n_slices-1] - density_depth_average[n_slices-2];
+
+              std::copy ( density_depth_average.begin(), density_depth_average.end(), padded_density_depth_average.begin() + 1 );
+
+              // Loop over the cells
+              for (const auto &cell : this->get_dof_handler().active_cell_iterators())
+                if (cell->is_locally_owned())
+                  {
+                    // Get the pressure, temperature and composition in the cell
+                    fe_values.reinit (cell);
+
+                    in.reinit (fe_values, cell, this->introspection(), this->get_solution());
+
+                    // Evaluate density at the cell center
+                    this->get_material_model().evaluate(in, out);
+                    const double density = out.densities[0];
+
+                    // Find the depth of the zeroth quadrature point in the cell and work out
+                    // the depth slice which has its center just above that point
+                    const double depth = this->get_geometry_model().depth(fe_values.quadrature_point(0));
+
+                    // Please note: static_cast<int> always truncates the double (i.e. 1.9 -> 1)
+                    // Here we subtract 0.5 to find the first slice with its center shallower than the cell,
+                    // and add 1 to take padding into account (-0.5 + 1.0 = +0.5)
+                    const double slice_depth = (depth*n_slices)/max_depth + 0.5;
+                    const unsigned int idx = static_cast<unsigned int>(slice_depth);
+                    const double fractional_slice = slice_depth - static_cast<double>(idx);
```

Comment:
I do not understand why you need the padding. The following should give you safe results within the bounds of `0` to `n_slices-1`:
```suggestion
                    // Please note: static_cast<int> always truncates the double (i.e. 1.9 -> 1)
                    const double slice_depth = (depth*n_slices)/max_depth;
                    const unsigned int idx = static_cast<unsigned int>(slice_depth);
                    const double fractional_slice = slice_depth - static_cast<double>(idx);
```

If there is a problem with this calculation then we should figure out why this problem exists.

Reply:
Add more explanation in the file.